cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/BufferingTable.java

```diff
@@ -249,46 +250,39 @@ public void startTx(Transaction tx) {
 
   @Override
   public Collection<byte[]> getTxChanges() {
-    switch (conflictLevel) {
-      case NONE:
-        return Collections.emptyList();
-      case ROW:
-        return getRowChanges();
-      case COLUMN:
-        return getColumnChanges();
-      default:
-        throw new RuntimeException("Unknown conflict detection level: " + conflictLevel);
+    if (ConflictDetection.NONE == conflictLevel) {
+      return Collections.emptyList();
     }
-  }
-
-  private Collection<byte[]> getRowChanges() {
-    // we resolve conflicts on row level of individual table
-    List<byte[]> changes = new ArrayList<byte[]>(buff.size());
-    for (byte[] changedRow : buff.keySet()) {
-      changes.add(Bytes.add(getNameAsTxChangePrefix(), changedRow));
+    Collection<byte[]> changes = new TreeSet<byte[]>(UnsignedBytes.lexicographicalComparator());
+    for (Map.Entry<byte[], NavigableMap<byte[], Update>> rowChange : buff.entrySet()) {
+      for (Map.Entry<byte[], Update> updateEntry : rowChange.getValue().entrySet()) {
+        byte[] changeKey = getChangeKey(rowChange.getKey(), updateEntry.getKey(), updateEntry.getValue());
+        if (changeKey != null) {
+          changes.add(Bytes.add(getNameAsTxChangePrefix(), changeKey));
```

Comment:
This incurs unnecessary overhead of byte copying etc. when conflict detection level is row. Because you need to do the Bytes.add() repeatedly for columns in the same row, then add to the set that already has that key. In case of ROW, you should only generate and add the change key once. I believe the old code was doing that. I also find the old code more intuitive.